defender/staged/downloaded/Compound_Gasless_Voting.js

```diff
@@ -0,0 +1,122 @@
+
+  // Test relay and check for balance
+  let relayAddress;
+  try {
+    relayAddress = await signer.getAddress();
+  } catch (error) {
+    console.error('Relay is not working, check if it is connected : ', error);
+    throw error;
+  }
+
+  const relayBalance = await provider.getBalance(relayAddress);
+  if (relayBalance <= 0) {
+    throw new Error('Insufficient funds for Relay, please deposit funds.')
+  }
+
+  // create an ethers.js Contract Object to interact with the on-chain governance smart contract
+  console.debug('Creating governanceContract');
+  const governanceContract = new ethers.Contract(
+    compoundGovernanceAddress,
+    compoundGovernanceAbi,
+    signer,
+  );
+
+  // create an ethers.js Contract Object to interact with the on-chain COMP smart contract
+  console.debug('Creating compContract');
+  const compTokenContract = new ethers.Contract(
+    compTokenAddress,
+    compTokenAbi,
+    signer,
+  );
+
+  // cast vote or delegate vote on-chain
+  const promises = body.map(async (votes) => {
+    if (votes.delegatee !== undefined) {
+      const {
+        address, delegatee, nonce, expiry, v, r, s,
+      } = votes;
+
+      // check if address has votes to delegate
+      const delegateCompBalance = await compTokenContract.balanceOf(address);
+      if (delegateCompBalance > 0) {
+        // delegate vote on-chain
+        console.debug(`Delgating votes from ${address} to ${delegatee}`);
+        try {
+          await compTokenContract.delegateBySig(delegatee, nonce, expiry, v, r, s);
+        } catch (error) {
+          console.error(error);
+        }
+      }
+    } else if (votes.support !== undefined) {
+      const {
+        address, proposalId, support, v, r, s,
+      } = votes;
+
+      // check if address has enough delegated votes to submit a vote
+      // check if proposal is active - only active proposals can be voted on
+      const addressDelegatedVotes = await compTokenContract.getCurrentVotes(address);
+      const state = await governanceContract.state(proposalId);
+      if (addressDelegatedVotes > 0 && state === 1) {
+        if (support === 0) {
+          console.debug(`Address ${address} is casting a vote in favor of proposal ID: ${proposalId}`);
+        } else if (support === 1) {
+          console.debug(`Address ${address} is casting a vote against proposal ID: ${proposalId}`);
+        } else if (support === 2) {
+          console.debug(`Address ${address} is casting a vote to abstain for proposal ID: ${proposalId}`);
+        }
+        // cast vote on-chain
+        try {
+          await governanceContract.castVoteBySig(proposalId, support, v, r, s);
+        } catch (error) {
+          console.error(error);
+        }
+      }
+    } else {
+      // error is not thrown to prevent one failed transaction from failing the rest in the batch
+      console.error('Action is not a vote or delgate-vote');
+    }
+  });
+  await Promise.all(promises);
+
```

Comment:
Piggybacking on @teasipper-ae 's error-handling comment. You can use the flag `let results = true` and anytime an error is encountered set the flag to false. Then at this final return statement `return results`. This would notify the webhook consumer that something did not work, and the autotask owner could look at the logs for this run for more details. 

Alternatively, the errors could be appended to the results and returned to the caller.

Reply:
I think this suggestion, combined with using `Promise.allSettled()`, is probably the right way to go.

Reply:
add this suggestion along with the `Promise.allSettled()`